### ISSUE TITLE
Race condition in PV controller storeObjectUpdate()

### ISSUE THREAD

**Issue by huww98** [kubernetes/kubernetes#124224]
### What happened?

PV controller (in KCM) has its own cache in addition to the informer cache, to store updates from itself. However, we've identified 2 possible race condition when invoking `storeObjectUpdate()` from multiple goroutines.

`storeObjectUpdate()` is currently invoked from `provisionClaimOperation()` and `syncUnboundClaim()`, which runs in 2 different goroutine. And I would expect more concurrent access for enhanced throughput in the future.

Race conditions:
1. internal update overwrites external delete. When a object is deleted from APIServer, an update request from controller itself may still in-flight. After the update returns, it tries to insert the object back to the cache, after deletion propagated from informer. Resulting in the object stay in cache forever. Subsequently causing, for example, PV stay in Bound state forever and cannot be released.
2. read-compare-write in `storeObjectUpdate()`. When invoked concurrently, an old version (defined by ResourceVersion) may overwrite new version.

### What did you expect to happen?

No race conditions. Internal cache stays consistent on concurrent access.

### How can we reproduce it (as minimally and precisely as possible)?

Hard to reproduce, this only happens with a small chance.

We observed condition 1 when we try to optimize the throughput of PV controller.

### Anything else we need to know?

To fix race condition 2, I propose: add locks around `storeObjectUpdate()`

To fix race condition 1, I propose: explicitly differentiate update and add event, do not allow add for internal update response.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by huww98** [comment]
/cc @jsafrane 
related #25881

**Comment by jsafrane** [comment]
>  1. internal update overwrites external delete. When a object is deleted from APIServer, an update request from controller itself may still in-flight. After the update returns, it tries to insert the object back to the cache, overwriting deletion propagated from informer. Resulting in the object staying in cache forever. Subsequently causing, for example, PV stay in Bound state forever and cannot be released.

@huww98 did you observe this race condition in a real cluster? I think it's mostly theoretical.

I could imagine something like a new parameter `storeObjectUpdate(..., store cache.Store, obj interface{}, force bool)` that would store the object when it's missing in the cache only when `force == true`. This would be used only in the initial sync or as reaction to informer event. Otherwise, if the object is missing (i.e. must have been deleted from the informer), then it's not added back to the cache. This is of course just an ugly hack on top of the cache. It would be the best to get rid of the whole cache instead. Smart ideas are welcome!

> 2. read-compare-write in storeObjectUpdate(). When invoked concurrently, an old version (defined by ResourceVersion) may overwrite new version. Resulting in unbounded cache stall.

This should get fixed by a periodic sync every 15 seconds. I know we should not depend on it and 15s is a bit aggressive, but it works here.

**Comment by huww98** [comment]
> did you observe this race condition in a real cluster? I think it's mostly theoretical.

Yes. I've seen this in our production cluster, which has some patches to increase the throughput of PV controller. We saw a PV stay in Bound for no reason. And after we created and deleted the referenced PVC once again, the PV was successfully released. We have seen this at least twice.
 
> I could imagine something like a new parameter `storeObjectUpdate(..., store cache.Store, obj interface{}, force bool)` that would store the object when it's missing in the cache only when `force == true`. This would be used only in the initial sync or as reaction to informer event. Otherwise, if the object is missing (i.e. must have been deleted from the informer), then it's not added back to the cache. This is of course just an ugly hack on top of the cache. It would be the best to get rid of the whole cache instead. Smart ideas are welcome!

Yes, this is exactly how I try to fix it in our internal branch.

We can also just use `Store.Add()` in informer add event handler and init. Then change `storeObjectUpdate()` to only support update.

I've experimented with another approach. Take the informer cache as the truth, and suppress the stall events. Use a `sync.Map` to maintain the latest generation of each UID, then ignore the update event of stall object, or do not consider stall objects when syncing. At least we cannot mistake deleted object as presented in this approach. But I need to check the generation when reading the cache. We can use `CompareAndSwap` to update the generation map, which seems to be more lightweight and elegant.

> This should get fixed by a periodic sync every 15 seconds. I know we should not depend on it and 15s is a bit aggressive, but it works here.

I agree, edited the issue to reflect this.

**Comment by jsafrane** [comment]
> I've experimented with another approach. Take the informer cache as the truth, and suppress the stall events. Use a sync.Map to maintain the latest generation of each UID, then ignore the update event of stall object, or do not consider stall objects when syncing. At least we cannot mistake deleted object as existing in this approach. But I need to check the generation when reading the cache. We can use CompareAndSwap to update the generation map, which seems to be more lightweight and elegant. We can also make good use of informer indexer in this approach.

Does `metdata.generation` even work on PVs and PVCs? I can see they're empty both of them.

**Comment by jsafrane** [comment]
`generation` usually changes when `.spec` changes, not `.status`. And PV controller saves `pv.spec`, `pvc.spec`, `pv.status` and `pvc.status` in one go and it would be great if we could ignore the `.status` updates. I am not sure it's possible with generation.
